node/tools/src/main.rs

```diff
@@ -1,66 +1,109 @@
 //! Main binary for the consensus node. It reads the configuration, initializes all parts of the node and
 //! manages communication between the actors. It is the main executable in this workspace.
 use anyhow::Context as _;
-use clap::Parser;
+use clap::{Args, Parser};
 use std::{fs, io::IsTerminal as _, path::PathBuf};
 use tracing::metadata::LevelFilter;
 use tracing_subscriber::{prelude::*, Registry};
 use vise_exporter::MetricsExporter;
 use zksync_concurrency::{ctx, scope};
-use zksync_consensus_tools::{decode_json, ConfigPaths, NodeAddr, RPCServer};
+use zksync_consensus_crypto::{Text, TextFmt};
+use zksync_consensus_roles::{node, validator};
+use zksync_consensus_tools::{decode_json, AppConfig, ConfigArgs, RPCServer};
 use zksync_protobuf::serde::Serde;
 
-/// Wrapper for Vec<NodeAddr>.
-#[derive(Debug, Clone)]
-struct NodeAddrs(Vec<Serde<NodeAddr>>);
-
-impl std::str::FromStr for NodeAddrs {
-    type Err = anyhow::Error;
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Ok(Self(decode_json(s)?))
-    }
-}
-
 /// Command-line application launching a node executor.
 #[derive(Debug, Parser)]
-struct Args {
-    /// Path to a validator key file. If set to an empty string, validator key will not be read
-    /// (i.e., a node will be initialized as a non-validator node).
-    #[arg(long, default_value = "./validator_key")]
-    validator_key: PathBuf,
-    /// Path to a JSON file with node configuration.
-    #[arg(long, default_value = "./config.json")]
-    config_file: PathBuf,
-    /// Path to a node key file.
-    #[arg(long, default_value = "./node_key")]
-    node_key: PathBuf,
+struct Cli {
+    /// Node configuration.
+    #[command(flatten)]
+    config_grp: Config,
+    /// Validator key.
+    #[command(flatten)]
+    validator_key_grp: ValidatorKey,
+    /// Node key definition.
+    #[command(flatten)]
+    node_key_grp: NodeKey,
     /// Path to the rocksdb database of the node.
     #[arg(long, default_value = "./database")]
     database: PathBuf,
     /// Port for the RPC server.
     #[arg(long)]
     rpc_port: Option<u16>,
-    /// IP address and key of the seed peers.
-    #[arg(long)]
-    add_gossip_static_outbound: Option<NodeAddrs>,
 }
 
-impl Args {
+/// Node key definitions:
+/// If both are present, node-key option has precedence.
+#[derive(Debug, Args)]
+#[group(required = false, multiple = false)]
```

Comment:
The best would be if these input options won't be mutually exclusive, but follow the pattern mentioned here: https://github.com/matter-labs/era-consensus/pull/62#discussion_r1493585570 (so that every field can be overridden independently).

You can see a similar example from a different project here: [config.go#L716](https://github.com/lightningnetwork/lnd/blob/fba6fdaf94be92063f19e8d6e92c7722278e156f/config.go#L716).

If it's too complicated we can leave it as is for now.

Reply:
Ok, I think it's good enough, because the the node running stand-alone isn't its main use case.

Reply:
Ok, I think it's good enough, because the the node running stand-alone isn't the main use case.